feature/45 - Implement *T tool

oobayly opened the pull request:
Resolves #45

The purpose of this pull request is to implement a *T (or equivalent) tool to measure distances, bearings on the scope.

Currently, it's initialted by holding down the Shift, although a scope command or button could be used to do the same.
* The tool will snap to the nearest aircraft or fix, or if none is found nearby it defaults to the cursor position
* Multiple ponts are supported, and can be an Aircraft, Fix, or Arbitrary point, and any combination of the above
* The path is updated as the mouse is moved to provide immediate vector information
* If the first point is an aircraft, the current ground speed will be used to estimate timings and turn radii

Isssues:
* There is currently no option to disable snapping. A 2nd keyboard modifier could be used.
* When drawing _"hairpin"_ paths, the tool will struggle to draw accurate arcs as the fillet (turn) radius may be too large.
* Tests are needed, as usual...

I've explain as much as possible in the comments, but I'm sure there'll be plenty of WFTs!

erikquinn commented:
> If the first point is an aircraft, the current ground speed will be used to estimate timings and turn radii

Actually, we should actively _not_ include any turn rounding. Straight lines from point to point to point only. If someone can provide a reliable reference showing that some places do round these lines, we can consider it as an option, though I may be inclined to fight it. :wink:

So for now, straight lines are perfect, which makes the work notably less difficult in terms of drawing.

oobayly commented:
> If someone can provide a reliable reference showing that some places do round these lines

The reason I went for arcing the lines was that I **vaguelly** remember seeing that in a VatSim ATC video, so I **assumed** that that was the way to go. Plus, it was a challenge to see if I could get it to work :grin:

To be fair, the all (but the first) arcs are cosmetic, as the distances are are calculated on vertex-to-vertex, but the errors would usually be fairly small relative to the length of the path. **I'll disable them for the time being**.

However, I'd argue that the initial turn could be a valuable feature as it calculates the turn to be made to fly to the point, and therefore the actual time to reach the point. Even if can be turned on in via the settings - just a thought...

I also noticed the build failed because of the _"dreaded prop"_, which actually raises something I'd considered, which was to make the `MeasureTool` a static object as it should never be used more than once.

![image](https://user-images.githubusercontent.com/3430117/67278998-14b8d100-f4c2-11e9-99fd-81c29832da1b.png)

n8rzz commented on src/assets/scripts/client/InputController.js:
```diff
@@ -206,6 +213,91 @@ export default class InputController {
+     * @param doReplaceLast {boolean} Indicates whether this will replace the last point
+     * @private
+     */
+    _addMeasurePoint(event, doReplaceLast = false) {
```
`doReplaceLast` -> `shouldReplaceLastPoint`

n8rzz commented on src/assets/scripts/client/InputController.js:
```diff
@@ -227,6 +319,13 @@ export default class InputController {
+
+        if (measureTool.isMeasuring && measureTool.hasStarted) {
+            this._addMeasurePoint(event, true);
+            return this;
```
missing empty line above `return`

n8rzz commented on src/assets/scripts/client/InputController.js:
```diff
@@ -502,6 +606,31 @@ export default class InputController {
+    _onKeyup(event) {
+        let { code } = event.originalEvent;
+
+        if (code === undefined) {
```
you should actually look for:
```js
if (typeof code === 'undefined') { }
```

or, if you want to use coercion:
```js
if (code == null) { }
```

oobayly replied:
Copy pasta from `_onKeyDown` :grin: I'll fix both in that case...

n8rzz commented on src/assets/scripts/client/InputController.js:
```diff
@@ -842,7 +988,11 @@ export default class InputController {
      */
     _onRightMousePress(event) {
-        this._markMousePressed(event, MOUSE_BUTTON_NAMES.RIGHT);
+        if (this.input.measureTool.isMeasuring) {
```
```js
if (!this.input.measureTool.isMeasuring) {
    this._markMousePressed(event, MOUSE_BUTTON_NAMES.RIGHT);

    return;
}

this._removeLastMeasurePoint();
```

oobayly replied:
I know it's much of a muchness, but would it be better to put the `_removeLastMeasurePoint` call inside the if? That way if more tests are ever added, the default behaviour will be the original `_markMousePressed`

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1217,6 +1218,144 @@ export default class CanvasController {
+
+        // This way the points are translated only once
+        while (leg != null) {
+            const p = CanvasStageModel.translatePostionModelToPreciseCanvasPosition(leg.midPoint);
```
does `p` stand for `position` or `point`?

oobayly replied:
Yeah, they're control positions in on the Canvas. Fixed

n8rzz commented on src/assets/scripts/client/measurement/MeasureLegModel.js:
```diff
@@ -0,0 +1,277 @@
+     * @type {number}
+     */
+    get radius() {
+        const response = this._radius;
```
what does `response` mean here?  feels like a strange word in relation to `radius` an measuring.

oobayly replied:
Don't judge me, but this is a side effect from my Basic days when I didn't like to name a local variable the same as the method it's declared in. :duck:
It sometimes leaks through...

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+     * @type {AircraftModel|null}
+     */
+    get aircraft() {
+        if (this.hasStarted && this._points[0] instanceof AircraftModel) {
```
please reverse this logic to return `null` early when we don't have what we need

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+        // The "last" point is actually the 2nd last item in _points
+        // This means the path will end at the cursor position when redrawn
+        const { length } = this._points;
+        if (length > 2) {
```
missing empty line above `if`

oobayly replied:
I don't know how you spot these, I always think I've done it.

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+     * @param startValue {AircraftModel|FixModel|array<number>}
+     */
+    startMeasuring(startValue) {
+        this._points = [startValue];
```
is `_points` an empty array before this assignment?  is so, you should use `.push()` here.  this is actually mutating `_points` instead of updating

oobayly replied:
Yes it should always be empty, but it was my way of making sure it was (which is a ~good~ bad way of ensuring bugs never manifest themselves)...

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+     * @param value {array<number>}
+     */
+    updateLastPoint(value) {
+        if (this.hasLegs) {
```
reverse the logic to return early:
```js
if (!this.hasLegs) {
    this.addPoint(value);

    return;
}

this._points[this._points.length - 1] = value;
```

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+     * @param bearing {number|null}
+     * @private
+     */
+    _getLabel(distance, duration, bearing = null) {
```
`_getLabel` -> `_buildLabel`

`get` has a very specific meaning and we aren't actually _*getting*_ anything here, we're calculating/building up.

oobayly replied:
Same goes for `getPathInfo` and `_getInitialTurnParameters` then...

n8rzz commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,363 @@
+     * Get the text label that displays the distance and
+     * optional duration and bearings
+     *
+     * @param distance {number}
```
what are the units of these numbers?

oobayly replied:
Ugh, one of my aerodynamics lecturers would dock a mark from coursework every time we left out a unit. Can't believe I left that out!

n8rzz commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -127,13 +148,7 @@ class FixCollection extends BaseCollection {
-        });
-
-        return _compact(realFixList);
+        return this._items.filter((fix) => fix.isRealFix);
```
😌

oobayly replied:
I had to google that emoji to make sure it was a good thing...

oobayly commented:
> This looks really, really good! Most of my feedback are nitpicks. The meat of this submission is really, really well put together!

Thanks, that's much appreciated.

> One thing I'd like to see a little more of is tests. Specifically, for the new Measure classes and/or anything thats new in a class that doesn't import jquery.

Definitely. I'd held off creating any as yet, as quite often a lot changes during the lifetime of a PR! For example, I've:
* In order to facilitate _"death to the prop"_, I've changed the `MeasureTool` to be a static object, although also I considered using the `EventBus` to pass information from the `InputController` to the `CanvasController`
* I've also added a _"Measure path style"_ setting that allows the user to control the render style.

erikquinn commented on src/assets/scripts/client/navigationLibrary/FixCollection.js:
```diff
@@ -101,6 +100,28 @@ class FixCollection extends BaseCollection {
+     * @param position {StaticPositionModel}
+     * @param allowRnavFixes {boolean} A flag indicating whether RNAV fixes should be used
+     */
+    getNearestFix(position, allowRnavFixes = false) {
```
`allowRnavFixes` --> `hiddenFixes`

(RNAV fixes are totally game, just not the fake ones we create to make certain procedures work, so I don't want to confuse readers of the code when they see that we aren't including "RNAV fixes", when really we _are_ 😉)

oobayly replied:
Hi hadn't considered that. :+1:

oobayly commented on src/assets/scripts/client/InputController.js:
```diff
@@ -342,15 +435,20 @@ export default class InputController {
         switch (code) {
+            case KEY_CODES.SHIFT_LEFT:
+            case KEY_CODES.SHIFT_RIGHT:
+                this._startMeasuring();
```
Funny, I started off with Ctrl, but switched to Shift after re-reading the original issue.

oobayly commented on src/assets/scripts/client/InputController.js:
```diff
@@ -828,10 +951,27 @@ export default class InputController {
+     */
+    _findClosestFixAndDistanceToMousePosition(x, y) {
+        return FixCollection.getNearestFix(
+            this._translatePointToKilometers(x, y)
```
My bad, I though I was replicating the behaviour of `AircraftController.aircraft_get_nearest`, obviously I wasn't...

oobayly commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1217,6 +1219,142 @@ export default class CanvasController {
+
+            // The angles calculated in the `MeasureTool` are magnetic, and have to be shifted CCW 90°
+            cc.arc(position.x, position.y, radius, entryAngle - Math.PI / 2, exitAngle - Math.PI / 2, !isRHT);
+        }
```
Yeah, this isn't entirely obvious. The way `MeasureTool.buildPathInfo` works is that it sets the radius to 0 if the corner shouldn't be arced (either from the style, or if the corner is too sharp to fillet). This has the same affect as using a simple lineTo. The reason it's done like that is twofold:
* Its simplified the draw logic by not having two loops, at the possible small expense of using the `arcTo` function (although I'd like to think it uses `lineTo` internally if the `radius === 0`)
* A path can have a mixture of arcs and points

oobayly commented on src/assets/scripts/client/constants/gameOptionConstants.js:
```diff
@@ -16,7 +16,8 @@ export const GAME_OPTION_NAMES = {
     MOUSE_CLICK_DRAG: 'mouseClickDrag',
-    RANGE_RINGS: 'rangeRings'
+    RANGE_RINGS: 'rangeRings',
+    MEASURE_TOOL_PATH: 'measureToolPath'
```
I remember this, I couldn't work out how the ordering is done, becuase `DRAW_ILS_DISTANCE_SEPARATOR` comes after `SOFT_CEILING`, so I thought they were somehow grouped by usage (which in hindsight is a little daft!).

erikquinn replied:
Definitely!

oobayly commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,513 @@
+        const { length } = this._points;
+
+        if (length > 2) {
+            this._points.splice(length - 2, 1);
```
Again, it's not very obvious. If I used `pop()`:
1. the last point would be removed as expected
2. but if the mouse isn't moved,  `updateLastPoint` would be not be called
3. meaning that the line will contine to end at the last corner, and not under the mouse
4. I found that behaviour a little clumsy.

With the current usage, this removes the last corner so the lines continues to end under the mouse, even if the mouse hasn't been moved. It just felt more intuitive that way, even though the code looks a tad strange.

erikquinn replied:
Oh god, I'm sorry. You even had a comment to this effect and I didn't notice it, or the fact that were splicing the second-to-last element, not the last. Disregard everything I said and carry on.

oobayly commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,513 @@
+            return null;
+        }
+
+        const { aircraft } = this;
```
Yup, the tool takes a `FixModel`, `AircraftModel` or `relativePosition` as any point. The `aircraft` property is simply an alias for the first point (providing it's an instance of `AircraftModel`).

This could possibly be seen as a feature. If the draw style is `Arc to next fix`, starting from a fix (or cursor position) and ending in an aircraft will result in a straight line. So you can still get "as the crow flies" distance.

oobayly commented on src/assets/scripts/client/InputController.js:
```diff
@@ -206,6 +211,88 @@ export default class InputController {
+        const currentMousePosition = CanvasStageModel.translateMousePositionToCanvasPosition(
+            event.pageX, event.pageY
+        );
+        const { x, y } = currentMousePosition;
```
Sorry, I'm not following. Are you saying you want to move the `findClosest` logic to the MeasureTool?

oobayly commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,513 @@
+    STRAIGHT: 0,
+    INITIAL_TURN: 1,
+    ARCED: 2
+};
```
I can see what you mean, but my only concern is that we'll be using string like `'straight'` in various places. Using the constants means they're only defined one.

I'm happy to go to string constants instead of bitwise operators, but wouldn't it make sense to keep `MEASURE_TOOL_STYLE` constant and import it into `gameOptionsConsts`?

``` js
            {
                displayLabel: 'Straight lines only',
                value: MEASURE_TOOL_STYLE.STRAIGHT
            },
```

erikquinn replied:
Fair point. And yes, that's a perfect solution then.

oobayly commented on src/assets/scripts/client/measurement/MeasureTool.js:
```diff
@@ -0,0 +1,513 @@
+         * @default MEASURE_TOOL_STYLE.STRAIGHT
+         * @private
+         */
+        this.style = MEASURE_TOOL_STYLE.STRAIGHT;
```
Will do.

oobayly commented:
This was easier that could have been expected, possibly because some part of me thought this might be requested.

1. Use Ctrl to start a new path
2. Start clicking (use Shift to snap)
3. Release Ctrl to end that path
4. Go to # 1
5. Press Esc to clear the tool

erikquinn commented:
@oobayly Is this working properly? It may be just me because I'm on a laptop with a bad trackpad, but I am having maybe a 5% success rate getting the lines to come up using ctrl. I thought it wasn't working at all until I had it work once or twice, but was very difficult to reliably create them. Just trying to figure out if it's me, or if you can confirm that something is up. 🤔

oobayly commented:
Strange, I had done most of the development for this on my Linux lapop, and I've just tried the deployment on the same machine and it works as expected.

However I also tried it on my Windows PC & TV with a wireless keyboard and it's struggling. Attaching a keydown event handler shows in the event being repeatedly called even though the key isn't being released. Can I ask what OS and browser version you're using?

erikquinn commented:
Twas on Win10, Chrome 77.0.3865.120

Updated to 78.0.3904.87 (latest) just now, and same result.

Maybe it has to do with it being a modifier? Not sure why it's behave differently than shift, but looking at changing to "on click, if Ctrl is currently depressed" rather than "on Ctrl keydown" if that's even how it's working (I no longer remember :grimacing:)

oobayly commented:
I've done all the dev work on Linux and hadn't noticed this on Windows. I've also noticed that  the keydown behaviour is different on  Chrome on Linux than it is on Windows 10. Using this website - https://howtodoinjava.com/jquery/jquery-difference-between-keypress-and-keydown-events/  and holding down Ctrl:
• On Linux, the keydown event is fired only once
• On Windows, the keydown event is fired repeatedly
It may be an OS thing (Edge does the same), fun, fun fun...

oobayly commented:
What was happening was that the repeated keypress events being fired were calling `startNewPath` every time, and that was ending the last path and starting a new one every event.

Instead of trying to handle different browser event behaviour, `MeasureTool.startNewPath` returns early if a path is already being measured.

I've just tried the deployment on my Windows box (after clearing the cache) and it works as expected (or at least the way I expect it to work)

erikquinn commented:
Looks like a travis issue 🤔

erikquinn commented:
Does seem to be working again for me on the laptop!

erikquinn commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1217,6 +1219,149 @@ export default class CanvasController {
+            const width = 150;
+
+            cc.fillRect(x, y, width, height);
+        });
```
Want to try to get the text box fill to always match the size of its content so it isn't taking up extra space that it doesn't need to (and thus blocking things behind it).

![image](https://user-images.githubusercontent.com/5103735/68353300-252e9580-00d7-11ea-99fc-b49838b98760.png)

oobayly replied:
I actually moved the colours to the themes, but stupidly forgot to push... 🤦‍♂️

I'll merge your changes and do the measurements (I actually removed the text box background for the default theme to march the style of the aircraft labels).

I also slightly changed the behaviour of start measuring to throw off it's called a second time so it can be tested for.

oobayly commented on src/assets/scripts/client/InputController.js:
```diff
@@ -1006,7 +1006,7 @@ export default class InputController {
     _onRightMousePress(event) {
         if (MeasureTool.isMeasuring) {
-            this._removeLastMeasurePoint();
+            this._removePreviousMeasurePoint();
```
That's much nicer behaviour, I wonder why I didn't think of that.

oobayly commented:
There were some changes I forgot to push up. They were mainly some code and test improvements, but also moved the colours to the themes. I used the colours you suggested for the Default theme, and used what I felt were optimal for the Classic theme.

And the labels are now sized to the text (that omission was down to laziness...)
